### DISCUSSION TITLE
Best way to allow a dynamic (unsafe) "run arbitrary SQL" page?

### DISCUSSION THREAD

**Discussion by shah** [sqlpage/SQLPage#183] [Q&A] [answered]
Hi - I'd like to offer my users of a SQLPage site a page called "query.sql" which would allow them to type in some arbitrary SQL (understanding that it will be "unsafe") in a form and when the form is submitted it would execute the query they provided and return the results as a table. 

If there's an example that already exists it seems my feeble search attempts have not been able to find it. 
thanks!

**Comment by lovasoa** [comment, marked as answer]
Hello !
  There is no such feature built into SQLPage itself, but you can use your database's own dynamic sql evaluation function to implement that:


1. SQLite: [SQLite Dynamic SQL with the *define* extension](https://github.com/nalgeon/sqlean/blob/main/docs/define.md)
2. PostgreSQL: [EXECUTE in PL/pgSQL](https://www.postgresql.org/docs/current/sql-execute.html) (you can wrap it in a function)
3. MySQL: [MySQL Dynamic SQL using EXECUTE](https://stackoverflow.com/questions/190776/how-to-have-dynamic-sql-in-mysql-stored-procedure)
4. Microsoft SQL Server: [Dynamic SQL in SQL Server using sp_executesql](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-executesql-transact-sql?view=sql-server-ver16)

see https://www.databasestar.com/dynamic-sql/
